custom_components/waste_collection_schedule/waste_collection_schedule/source/c_trace_buergerportal.py

```diff
@@ -0,0 +1,233 @@
+import contextlib
+import re
+from base64 import standard_b64decode
+from datetime import datetime
+from typing import List, Optional, TypedDict, Union
+
+import requests
+from waste_collection_schedule import Collection  # type: ignore[attr-defined]
+
+TITLE = "C-Trace Bürgerportal"
+URL = "https://www.c-trace.de"
+DESCRIPTION = "Source for waste collection in multiple service areas."
+TEST_CASES = {
+    # "Cochem-Zell": {
+    #     "url": "https://buerger-portal-cochemzell.azurewebsites.net/",
+    #     "district": "Bullay",
+    #     "street": "Layenweg",
+    #     "number": 3,
+    # },
+    "Alb-Donau-Kreis": {
+        "url": "https://buerger-portal-albdonaukreisabfallwirtschaft.azurewebsites.net",
+        "district": "Blaubeuren",
+        "street": "Alberstraße",
+        "number": 3,
+    },
+    "Biedenkopf MZV": {
+        "url": "https://biedenkopfmzv.buergerportal.digital/",
+        "district": "Biedenkopf",
+        "subdistrict": "Breidenstein",
+        "street": "Auf dem Hammer",
+        "number": 1,
+    },
+}
+ICONS = {
+    "mobil": "mdi:truck",
+    "bio": "mdi:leaf",
+    "papier": "mdi:package-variant",
+    "verpackung": "mdi:recycle",
+    "gelb": "mdi:recycle",
+    "rest": "mdi:trash-can",
+    "gruen": "mdi:forest",
+    "grün": "mdi:forest",
+    "baum": "mdi:forest",
+    "schnitt": "mdi:forest",
+    "schad": "mdi:biohazard",
+}
+API_HEADERS = {
+    "Accept": "application/json, text/plain;q=0.5",
+    "Cache-Control": "no-cache",
+}
+
+
+def quote_none(value: Optional[str]) -> str:
+    if value is None:
+        return "null"
+
+    return f"'{value}'"
+
+
+class Source:
+    def __init__(
+        self,
+        url: str,
+        district: str,
+        street: str,
+        number: Union[int, str, None] = None,
+        subdistrict: Optional[str] = None,
+    ):
+        self.api_url = f"{url.removesuffix('/')}/api"
+        self.district = district
+        self.subdistrict = subdistrict
+        self.street = street
+        self.number = number
+
+    def fetch(self):
+        session = requests.session()
+        session.headers.update(API_HEADERS)
+
+        year = datetime.now().year
+        entries: list[Collection] = []
+
+        district_id = self.fetch_district_id(session)
+        street_id = self.fetch_street_id(session, district_id)
+        # Eventually verify house number in the future
+
+        params = {
+            "$expand": "Abfuhrplan,Abfuhrplan/GefaesstarifArt/Abfallart,Abfuhrplan/GefaesstarifArt/Volumen",
+            "$orderby": "Abfuhrplan/GefaesstarifArt/Abfallart/Name,Abfuhrplan/GefaesstarifArt/Volumen/VolumenWert",
+            "orteId": district_id,
+            "strassenId": street_id,
+            "jahr": year,
```

Comment:
Do your really need year as argument? Some service providers return all available data is you don't specify a date. If not (= year is required), could you please fetch the data for the current year and in December also the data for the next year?

Reply:
Yes, we need the year as an argument. The API endpoint `AbfuhrtermineAbJahr` returns all upcoming waste collections starting with a certain `year`. In my tests, I sometimes had the scenario that only collections in 2023 were returned if leaving this parameter out. If given the year 2022, then we will also receive all collections from 2023: The parameter `year` is only meant to specify the *starting point*.